test/common/upstream/health_checker_impl_test.cc

```diff
@@ -2467,6 +2467,121 @@ TEST_F(HttpHealthCheckerImplTest, Http2ClusterUseHttp2CodecClient) {
+  EXPECT_CALL(*default_socket_factory, createTransportSocket(_)).Times(0);
+  auto transport_socket_match = std::make_unique<NiceMock<Upstream::MockTransportSocketMatcher>>(
+      std::move(default_socket_factory));
+
+  auto metadata = TestUtility::parseYaml<envoy::config::core::v3::Metadata>(
+      R"EOF(
+    filter_metadata:
+      envoy.transport_socket_match:
+        key: value
+  )EOF");
+
+  Stats::IsolatedStoreImpl stats_store;
+  auto health_transport_socket_stats = TransportSocketMatchStats{
+      ALL_TRANSPORT_SOCKET_MATCH_STATS(POOL_COUNTER_PREFIX(stats_store, "test"))};
+  auto health_check_only_socket_factory =
+      std::make_unique<NiceMock<Network::MockTransportSocketFactory>>();
+
+  // We expect resolve() to be called twice, once for endpoint socket matching (with no metadata in
+  // this test) and once for health check socket matching. In the latter we expect metadata that
+  // matches the above object.
+  EXPECT_CALL(*transport_socket_match, resolve(nullptr));
+  EXPECT_CALL(*transport_socket_match, resolve(MetadataEq(metadata)))
+      .WillOnce(Return(TransportSocketMatcher::MatchData(
+          *health_check_only_socket_factory, health_transport_socket_stats, "health_check_only")));
+  // The health_check_only_socket_factory should be used to create a transport socket for the health
+  // check connection.
+  EXPECT_CALL(*health_check_only_socket_factory, createTransportSocket(_));
+
+  cluster_->info_->transport_socket_matcher_ = std::move(transport_socket_match);
+
+  health_checker_ = std::make_shared<TestHttpHealthCheckerImpl>(
+      *cluster_, parseHealthCheckFromV2Yaml(yaml), dispatcher_, runtime_, random_,
+      HealthCheckEventLoggerPtr(event_logger_));
+
+  cluster_->prioritySet().getMockHostSet(0)->hosts_ = {
+      makeTestHost(cluster_->info_, "tcp://127.0.0.1:80")};
+  cluster_->info_->stats().upstream_cx_total_.inc();
+  expectSessionCreate();
+  expectStreamCreate(0);
+  EXPECT_CALL(*test_sessions_[0]->timeout_timer_, enableTimer(_, _));
+  health_checker_->start();
+  EXPECT_EQ(health_transport_socket_stats.total_match_count_.value(), 1);
+}
+
+TEST_F(HttpHealthCheckerImplTest, NoTransportSocketMatchCriteria) {
+  const std::string host = "fake_cluster";
+  const std::string path = "/healthcheck";
+  const std::string yaml = R"EOF(
+    timeout: 1s
+    interval: 1s
+    no_traffic_interval: 1s
+    interval_jitter_percent: 40
+    unhealthy_threshold: 2
+    healthy_threshold: 2
+    http_health_check:
+      service_name_matcher:
+        prefix: locations
+      path: /healthcheck
+    )EOF";
+
+  auto default_socket_factory = std::make_unique<NiceMock<Network::MockTransportSocketFactory>>();
+  // The default_socket_factory should be used to create a transport socket for the health check
+  // connection.
+  EXPECT_CALL(*default_socket_factory, createTransportSocket(_));
+  auto transport_socket_match = std::make_unique<NiceMock<Upstream::MockTransportSocketMatcher>>(
```

Comment:
In general it might be nice to make all of these transport socket factory/matcher mocks strict so that it'll force you to list all the interactions

Reply:
When _should_ you use NickMock?

Reply:
Its useful when adding explicit expectations clutter the tests unnecessarily, e.g. a lot of mocks will default to returning a NiceMock version when called so that you aren't responsible for adding expectations to dozens of test classes that gets introduced when you use a single mock

Generally what I follow is to use strict mocks for the classes directly related to the test, and NiceMock for ancillary mocks that had to be included in order to get things compiling